CONTRIBUTING.md

```diff
@@ -57,3 +57,325 @@ If the changes in your PR do not conflict with any of the existing code in the p
+1. Implement all the methods for `ConfigEntryResource` in the `_types.go` file. If using goland you can
+   automatically stub out all the methods by using Code -> Generate -> IngressGateway -> ConfigEntryResource.
+1. Use existing implementations of other types to implement the methods. We have to
+   copy their code because we can't use a common struct that implements the methods
+   because that messes up the CRD code generation. 
+   
+   You should be able to follow the other "normal" types. The non-normal types
+   are `ServiceIntention` and `ProxyDefault` because they have special behaviour
+   around being global or their spec not matching up with Consul's directly.
+1. When you get to `ToConsul` and `Validate` you'll need to actually think
+   about the implementation instead of copy/pasting and doing a simple replace.
+1. For `ToConsul`, the pattern we follow is to implement `toConsul()` methods
+   on each sub-struct. You can see this pattern in the existing types.
+1. For `Validate`, we again follow the pattern of implementing the method on
+   each sub-struct. You'll need to read the Consul documentation to understand
+   what validation needs to be done.
+   
+   Things to keep in mind:
+   1. Re-use the `sliceContains` and `notInSliceMessage` helper methods where applicable.
+   1. If the invalid field is an entire struct, encode as json (look for `asJSON` for an example).
+   1. `validateNamespaces` should be a separate method.
+   1. If the field can have a `nil` pointer, check for that, e.g.
+        ```go
+        func (in *ServiceRouteHTTPMatchHeader) validate(path *field.Path) *field.Error {
+            if in == nil {
+                return nil
+            }
+        ```
+
+### Spec Tests
+1. Create a test file, e.g. `ingressgateway_types_test.go`.
+1. Copy the tests for the `ConfigEntryResource` methods from another type and search and replace.
+   Only the tests for `ToConsul()`, `Validate()` and `MatchesConsul()` need to
+   be implemented without copying.
+1. The test for `MatchesConsul` will look like:
+    ```go
+    func TestIngressGateway_MatchesConsul(t *testing.T) {
+        cases := map[string]struct {
+            Ours    IngressGateway
+            Theirs  capi.ConfigEntry
+            Matches bool
+        }{
+            "empty fields matches": {
+            "all fields set matches": {
+            "different types does not match": {
+
+        }
+
+        for name, c := range cases {
+            t.Run(name, func(t *testing.T) {
+                require.Equal(t, c.Matches, c.Ours.MatchesConsul(c.Theirs))
+            })
+        }
+    }
+    ```
+1. The test for `ToConsul` will re-use the same struct as for the previous test cases
+   `empty field matches` and `all fields set matches` but renamed to `empty fields` and `every field set`
+   and with any `Namespace` set on the Consul struct removed and `CreateIndex`/`ModifyIndex` removed
+   since the `ToConsul` method won't set those.
```

Comment:
This is a mouthful, wondering if it can be rephrased with bullets for keep/remove?

Reply:
👍 split it up